ext/drreg/drreg.c

```diff
@@ -1985,13 +1988,30 @@ is_aflags_spill(instr_t *inst)
 #endif
 }
 
+static reg_id_t
+aflags_spill_get_reg(instr_t *inst)
+{
+#if !(defined(X86) || defined(AARCHXX))
+    ASSERT(false, "aflags_spill_get_reg");
+#endif
+    return IF_X86_ELSE(DR_REG_XAX, opnd_get_reg(instr_get_dst(inst, 0)));
+}
+
 static inline bool
 is_aflags_restore(instr_t *inst)
 {
 #if defined(X86)
     return instr_get_opcode(inst) == OP_sahf;
-#elif defined(AARCHXX)
-    return instr_get_opcode(inst) == OP_msr;
+#elif defined(ARM)
+    return (instr_get_opcode(inst) == OP_msr && instr_num_dsts(inst) == 1 &&
+            opnd_is_reg(instr_get_dst(inst, 0)) &&
+            opnd_get_reg(instr_get_dst(inst, 0)) == DR_REG_CPSR &&
+            instr_num_srcs(inst) == 2 && opnd_is_reg(instr_get_src(inst, 1)));
```

Comment:
Out of curiosity: what's the other src besides the reg that holds the flags to be restored to cpsr?

Reply:
It's a 5-bit constant. The highest bit (bit 4) selects SPSR instead of CPSR, and the other four bits indicate which bytes of the destination register are written to. So DynamoRIO's IR is perhaps not ideal. One could perhaps add a check here that bit 3 of the mask is set because it's the top byte of CPSR that contains NZCV (and Q). It's UNPREDICTABLE if you try to write to SPSR in user mode.

Reply:
I've also added code to check the first src opnd.